File: benchmarks/legacy/binary_tree.c
/* binary_tree.c — C reference implementation for the Arukellt binary_tree
 * benchmark.  Matches benchmarks/binary_tree.ark: count nodes (depth 20).
 * Compile: cc -O2 -o binary_tree binary_tree.c
 * Run:     ./binary_tree
 */
#include <stdio.h>

static int count_nodes(int depth) {
    if (depth == 0)
        return 1;
    return 1 + count_nodes(depth - 1) + count_nodes(depth - 1);
}

int main(void) {
    printf("%d\n", count_nodes(20));
    return 0;
}

File: benchmarks/legacy/binary_tree.rs
// binary_tree.rs — Rust reference implementation for the Arukellt
// binary_tree benchmark.  Matches benchmarks/binary_tree.ark: count nodes
// (depth 20).
// Compile: rustc -O -o binary_tree binary_tree.rs
// Run:     ./binary_tree

fn count_nodes(depth: i32) -> i32 {
    if depth == 0 {
        return 1;
    }
    1 + count_nodes(depth - 1) + count_nodes(depth - 1)
}

fn main() {
    println!("{}", count_nodes(20));
}
